zebra/zserv.c

```diff
@@ -220,176 +157,163 @@ static void zserv_log_message(const char *errmsg, struct stream *msg,
+ * instead of locking and unlocking each time we want to pop a single message
+ * off the queue. The same thing could arguably be accomplished faster by
+ * allowing the main thread to write directly into the buffer instead of
+ * enqueuing packets onto an intermediary queue, but the intermediary queue
+ * allows us to expose information about input and output queues to the user in
+ * terms of number of packets rather than size of data.
  */
 static int zserv_write(struct thread *thread)
 {
 	struct zserv *client = THREAD_ARG(thread);
 	struct stream *msg;
-	int writerv;
-
-	if (client->t_suicide)
-		return -1;
+	uint32_t wcmd;
+	struct stream_fifo *cache;
 
-	if (client->is_synchronous)
+	if (atomic_load_explicit(&client->dead, memory_order_seq_cst))
 		return 0;
 
-	msg = stream_fifo_pop(client->obuf_fifo);
-	stream_set_getp(msg, 0);
-	client->last_write_cmd = stream_getw_from(msg, 6);
-
-	writerv = buffer_write(client->wb, client->sock, STREAM_DATA(msg),
-			       stream_get_endp(msg));
-
-	stream_free(msg);
-
-	switch (writerv) {
+	/* If we have any data pending, try to flush it first */
+	switch (buffer_flush_all(client->wb, client->sock)) {
 	case BUFFER_ERROR:
-		zlog_warn(
-			"%s: buffer_write failed to zserv client fd %d, closing",
-			__func__, client->sock);
-		/*
-		 * Schedule a delayed close since many of the functions that
-		 * call this one do not check the return code. They do not
-		 * allow for the possibility that an I/O error may have caused
-		 * the client to be deleted.
-		 */
-		client->t_suicide = NULL;
-		thread_add_event(zebrad.master, zserv_delayed_close, client, 0,
-				 &client->t_suicide);
-		return -1;
-	case BUFFER_EMPTY:
-		THREAD_OFF(client->t_write);
-		break;
+		goto zwrite_fail;
 	case BUFFER_PENDING:
-		thread_add_write(zebrad.master, zserv_flush_data, client,
-				 client->sock, &client->t_write);
+		atomic_store_explicit(&client->last_write_time,
+				      (uint32_t)monotime(NULL),
+				      memory_order_relaxed);
+		zserv_client_event(client, ZSERV_CLIENT_WRITE);
+		return 0;
+	case BUFFER_EMPTY:
 		break;
 	}
 
-	if (client->obuf_fifo->count)
-		zebra_event(client, ZEBRA_WRITE);
-
-	client->last_write_time = monotime(NULL);
-	return 0;
-}
-
-#if defined(HANDLE_ZAPI_FUZZING)
-static void zserv_write_incoming(struct stream *orig, uint16_t command)
-{
-	char fname[MAXPATHLEN];
-	struct stream *copy;
-	int fd = -1;
+	cache = stream_fifo_new();
 
-	copy = stream_dup(orig);
-	stream_set_getp(copy, 0);
+	pthread_mutex_lock(&client->obuf_mtx);
+	{
+		while (client->obuf_fifo->head)
+			stream_fifo_push(cache,
+					 stream_fifo_pop(client->obuf_fifo));
+	}
+	pthread_mutex_unlock(&client->obuf_mtx);
 
-	zserv_privs.change(ZPRIVS_RAISE);
-	snprintf(fname, MAXPATHLEN, "%s/%u", DAEMON_VTY_DIR, command);
-	fd = open(fname, O_CREAT | O_WRONLY | O_EXCL, 0644);
-	stream_flush(copy, fd);
-	close(fd);
-	zserv_privs.change(ZPRIVS_LOWER);
-	stream_free(copy);
-}
-#endif
+	if (cache->tail) {
```

Comment:
Don't love poking around inside the fifo struct - is there a "stream_fifo_tail()" api?

Reply:
Nope, can add